packages/adapter/lib/adapter/adapter.js

```diff
@@ -1362,40 +1361,58 @@ function Adapter(options) {
             return;
         }
 
-        timerId = (timerId + 1) % 0xffffffff;
+        if (stopInProgress) {
+            this.log.warn(`setTimeout called, but adapter is shutting down"`);
+            return;
+        }
 
-        const _cb = (function (id, _cb) {
-            return function (...args) {
-                delete timers[id];
-                _cb(...args);
-            };
-        })(timerId, cb);
+        const id = setTimeout.call(null, () => {
+            timers.delete(id);
+            cb(...args);
+        }, timeout);
+        timers.add(id);
 
-        timers[timerId] = setTimeout.call(null, _cb, timeout, ...args);
-        return timerId;
+        return id;
     };
 
     /**
      * Same as clearTimeout
      * but it check the running timers on unload
      *
-     * It returns promise if no callback is provided.
      * @param {number} id - timer id
      */
     this.clearTimeout = id => {
-        if (!timers[id]) {
-            logger.warn(`${this.namespaceLog} Clear yet terminated timer ${id}`);
+        if (!timers.has(id)) {
+            logger.warn(`${this.namespaceLog} Clear already terminated timer ${id}`);
         } else {
-            clearTimeout(timers[id]);
-            delete timers[id];
+            clearTimeout(id);
+            timers.delete(id);
         }
     };
 
+    /**
+     * delays the fullfillment of the promise the amount of time.
+     * it will not fullfill during and after adapter shutdown
+     *
+     * @param {number} timeout - timeout in milliseconds
+     * @returns {Promise<void>} promise when timeout is over
+     */
+    this.delay = timeout => {
+        return new Promise(resolve => {
```

Comment:
Should we add the "if (stopInProgress) log-warn + return" guard here too? (but return the promise that never will resolve?

Reply:
if you do not allow timeout you should also not allow this, its just a promisification of `setTimeout`

Reply:
Yes but we need to do differently ... add a return in first place will simply make the sleep ignored and next code is executed ... so we need to make sure it is execvuted, promise returned but never resolved, so in fact you mean "yes"

Reply:
```suggestion
        return new Promise(resolve => {
            if (stopInProgress) {
                this.log.warn(`delay called, but adapter is shutting down`);
                return;
            }
```

Reply:
kind of that way I would say